presto-main/src/main/java/com/facebook/presto/operator/SimpleArrayAllocator.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.presto.operator;
+
+import com.facebook.presto.spi.block.ArrayAllocator;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+import java.util.ArrayDeque;
+import java.util.Deque;
+
+import static io.airlift.slice.SizeOf.sizeOf;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * A simple {@link ArrayAllocator} that caches returned arrays.  When retrieving an array which is
+ * cached, this implementation will first discard all smaller arrays encountered to ensure amortized
+ * constant time access to the appropriately sized array.
+ */
+@NotThreadSafe
+class SimpleArrayAllocator
+        implements ArrayAllocator
+{
+    private static final int DEFAULT_MAX_OUTSTANDING = 1000;
+    private final int maxOutstandingArrays;
+
+    private final Deque<int[]> intArrays = new ArrayDeque<>();
+    private int outstandingArrays;
+    private long estimatedSizeInBytes;
+
+    public SimpleArrayAllocator()
+    {
+        this(DEFAULT_MAX_OUTSTANDING);
+    }
+
+    public SimpleArrayAllocator(int maxOutstandingArrays)
+    {
+        if (maxOutstandingArrays <= 0) {
+            throw new IllegalArgumentException("maxOutstandingArrays must be positive");
+        }
+        this.maxOutstandingArrays = maxOutstandingArrays;
+    }
+
+    @Override
+    public int[] borrowIntArray(int positionCount)
+    {
+        checkState(outstandingArrays < maxOutstandingArrays, "Requested too many arrays: %s", outstandingArrays);
+        outstandingArrays++;
+        while (!intArrays.isEmpty() && intArrays.peek().length < positionCount) {
+            int[] popped = intArrays.pop();
+            estimatedSizeInBytes -= sizeOf(popped);
+        }
+        if (intArrays.isEmpty()) {
+            int[] array = new int[positionCount];
+            estimatedSizeInBytes += sizeOf(array);
+            return array;
+        }
+        return intArrays.pop();
+    }
+
+    @Override
+    public void returnArray(int[] array)
+    {
+        requireNonNull(array, "array is null");
+        checkState(outstandingArrays > 0, "Returned more arrays than leased");
+        outstandingArrays--;
+        intArrays.push(array);
```

Comment:
This method doesn't check that the array being returned is the same array that was borrowed, e.g. `returnArray(new int[10])` works just fine. How are you thinking about this? Are there significant performance implications of maintaining the set of pointers to the borrowed arrays to allow for tighter checking here?

Reply:
I could add a `newSetFromMap(new IdentityHashMap<>());` and collapse the `maxOutstandingArrays` into the `size()` method of that.